Standards/scs-0213-v1-k8s-nodes-anti-affinity.md

```diff
@@ -0,0 +1,123 @@
+
+## Design considerations
+
+SCS plans to require a Hard Anti Affinity and/or Redundancy for the control plane and
+a Soft Anti Affinity for workers. This means, that Anti Affinity would be required for
+the control planes and their pods and only optional (but encouraged) for workers.
+
+In order to achieve the goals for these components, meaning availability and fault tolerance
+for the control plane, in order to achieve an outage resistant cluster, and the availability
+promise given with Anti Affinity for pods on the worker nodes, a separation of nodes
+on the hardware or hypervisor level would need to be achieved. Since running multiple
+hypervisors on the same machine is in most cases not feasible and this approach
+wouldn't change availability for outages of individual machines, this should not be recommended.
+
+For the control plane, a reference to the "Kubernetes High Availability" [1]
+can be useful, since it provides two ways to set up a highly available cluster.
+Both approaches are very similar. The "Stacked Control Plane" requires less infrastructure,
+but also runs the risk of failed coupling, where if one node fails, the redundancy could be
+compromised due to the loss of a complete control plane instance.
+The "External ETCD" solves this problem, but also requires double the infrastructure, due
+to the externally incorporated etcd clusters.
+
+This also shows, that the wording "anti affinity" as used with Kubernetes pods is probably
+slightly misleading in the context of a Kubernetes control plane. It may consist of multiple
+pods with individual tasks, but distributing them over different nodes through Anti Affinity
+would probably still cascade the whole control plane into a complete failure, if one of
+the used nodes goes down. It could be possible to replicate specific important pods and
+assign them to different nodes, but at this point, a redundant setup like presented in [1] could be used.
+So Anti Affinity in this context probably means more like distribution over multiple
+physical machines, which needs to be planned beforehand on the machine/server level.
+
+Therefore would it be preferred for the control plane to use a redundant setup, which
+is separated over different physical machines, meaning at least half of the control plane
+nodes runs on a different physical machine as the rest.
+
+For worker nodes, the whole idea of Anti Affinity is more of a preferable situation.
+The nodes themselves should at best be distributed over different machines, but this
+is not a requirement, especially since smaller providers wouldn't have the capacity to
+provide enough machines for these distributed Kubernetes clusters. Since customers that
+use the Affinity feature would be especially interested in the worker nodes that host
+their workloads, it should be achieved that a good labeling system is provided for the
+nodes in order to see if two nodes are hosted on the same machine.
```

Comment:
IMO, in the case of workload pods it is worth also mentioning the k8s concept called [Pod Topology Spread Constraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/).

This concept is very close to the anti-affinity feature, but there are some differences, nicely explained e.g. [here](https://stackoverflow.com/questions/73157345/kubernetes-spread-pods-across-nodes-using-podantiaffinity-vs-topologyspreadconst)

Reply:
Thats a nice concept and good to know it exists. I don't think its that relevant in the end for this DR/Standard, but I mentioned it nonetheless.